utils/core/src/serde/byte_reader.rs

```diff
@@ -194,10 +199,432 @@ pub trait ByteReader {
     }
 }
 
+// STANDARD LIBRARY ADAPTER
+// ================================================================================================
+
+/// An adapter of [ByteReader] to any type that implements [std::io::Read]
+///
+/// In particular, this covers things like [std::fs::File], standard input, etc.
+#[cfg(feature = "std")]
+pub struct ReadAdapter<'a> {
+    // NOTE: The [ByteReader] trait does not currently support reader implementations that require
+    // mutation during `peek_u8`, `has_more_bytes`, and `check_eor`. These (or equivalent)
+    // operations on the standard library [std::io::BufRead] trait require a mutable reference, as
+    // it may be necessary to read from the underlying input to implement them.
+    //
+    // To handle this, we wrap the underlying reader in an [RefCell], this allows us to mutate the
+    // reader if necessary during a call to one of the above-mentioned trait methods, without
+    // sacrificing safety - at the cost of enforcing Rust's borrowing semantics dynamically.
+    //
+    // This should not be a problem in practice, except in the case where `read_slice` is called,
+    // and the reference returned is from `reader` directly, rather than `buf`. If a call to one
+    // of the above-mentioned methods is made while that reference is live, and we attempt to read
+    // from `reader`, a panic will occur.
+    //
+    // Ultimately, this should be addressed by making the [ByteReader] trait align with the standard
+    // library I/O traits, so this is a temporary solution.
+    reader: RefCell<std::io::BufReader<&'a mut dyn std::io::Read>>,
```

Comment:
It would certainly allow us to do away with the `RefCell` hack, and ensure that all uses of the `ByteReader` API are valid (i.e. there aren't hidden edge cases where a dynamic borrow check could fail).

I _think_ it would also allow us to implement `ByteReader` for `std::io::BufReader` directly, and get rid of `ReadAdapter` entirely, which would be nice for sure. There is one API that would be nice to have, but is still unstable (`has_data_left`), but we can implement an alternative as long as we have a mutable reference on hand. There is also the question of fallible reads with `std::io::BufReader`, i.e. if you attempt to read N bytes, but only read N-1 are available, but the internal buffer of the `BufReader` is say, `N / 2`, then we can't put the extra bytes back into the reader for a subsequent call - in other words, the data is lost.

That said, it's an edge case, and I'm not sure what the guarantees are of `ByteReader` with regard to read errors and subsequent reads following the error, but it seems to be implied that you could attempt to read the same data multiple times until you succeed, and `ReadAdapter` implements the necessary bookkeeping to support that.